# TP3 - Exercise 2
## Mateus Ribeiro da Silva

In [28]:
#Importing some libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
from scipy.stats import norm
from sklearn.utils import shuffle
from sklearn import neighbors
from sklearn.preprocessing import LabelEncoder
from sklearn import metrics
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV


# Create color maps for 3-class classification problem, as with iris
cmap_light = ListedColormap(['#FFAAAA', '#AAFFAA', '#AAAAFF'])
cmap_bold = ListedColormap(['#FF0000', '#00FF00', '#0000FF'])

In [29]:
def dataset_fusion(A, B):

    """
    This function merge two datasets and sort them based on the value column
    """

    dataset = pd.merge(A, B, how='outer')
    dataset = dataset.sample(frac=1).reset_index(drop=True)

    return dataset 

In [30]:
#Creating datasets
np.random.seed(25)

dataset_A = pd.DataFrame({'Value 1': np.random.normal(loc=1, scale=1, size=(4000,1)).flatten().tolist(),
                          'Value 2': np.random.normal(loc=21, scale=0.5, size=(4000,1)).flatten().tolist(),
                          'Dataset': ['A' for i in range(4000)]})
dataset_B = pd.DataFrame({'Value 1': np.random.normal(loc=2, scale=0.5, size=(4000,1)).flatten().tolist(),
                          'Value 2': np.random.normal(loc=19, scale=1, size=(4000,1)).flatten().tolist(),
                          'Dataset': ['B' for i in range(4000)]})

full_dataset = dataset_fusion(dataset_A, dataset_B)

In [31]:
#Transform the targets in variables that can be used in classifier training
le = LabelEncoder()
full_dataset['Dataset'] = le.fit_transform(full_dataset['Dataset'])

#Separate varibles from targets
X, y = full_dataset[['Value 1', 'Value 2']], full_dataset['Dataset']

In [32]:
#KNN classifier instance with k=10
knn = neighbors.KNeighborsClassifier(n_neighbors=10)

#Train model using cross validation split dataset over 10 groups and obtain the score from each split.
cross_val_scores = cross_val_score(knn, X, y, cv=10)

#Compute the average accuracy from 10 cross validations
print('Average Accuracy:', np.mean(cross_val_scores))


Average Accuracy: 0.9297500000000001


### Using cross validation which split out dataset over 10 groups, we got approximately 92.98% of accuracy. Such result is pretty good, however, we can try another values of k and find the best one. Let's use the GridSerach tool.

In [33]:
#Instance a new knn model
knn_new = neighbors.KNeighborsClassifier()

#Create grid dictionary containing all hyperparameters we would like to find. For KNN, we only will consider 'k' nearest neighbors.
param_grid = {'n_neighbors': [k for k in range(1, 100)]}

#Instance GridSearch to test all k using 10 cross validation
knn_grid = GridSearchCV(knn_new, param_grid, cv=10)

#Fit model
knn_grid.fit(X, y)

GridSearchCV(cv=10, estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12,
                                         13, 14, 15, 16, 17, 18, 19, 20, 21, 22,
                                         23, 24, 25, 26, 27, 28, 29, 30, ...]})

In [34]:
#Find the best performance
knn_grid.best_params_

{'n_neighbors': 69}

### The best k we could consider, according to purposed GridSearch, would be 69 nearest neighbors.

In [35]:
#Find the best score performance
knn_grid.best_score_

0.9347500000000002

### Considering k=69 and using 10 cross validation we reach out 93.48% of accuracy. So, GridSearchCV is a great tool for tuning classifier model, because it's very pratical, in comparison with all fullfilled effort in exercise 1.